secret/storage.go

```diff
@@ -0,0 +1,190 @@
+package secret
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+
+	"github.com/influxdata/influxdb/v2"
+	"github.com/influxdata/influxdb/v2/kv"
+)
+
+var secretBucket = []byte("secretsv1")
+
+// Storage is a store translation layer between the data storage unit and the
+// service layer.
+type Storage struct {
+	store kv.Store
+}
+
+// NewStore creates a new storage system
+func NewStore(s kv.Store) (*Storage, error) {
+	err := s.Update(context.Background(), func(tx kv.Tx) error {
+		if _, err := tx.Bucket(secretBucket); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Storage{s}, nil
+}
+
+func (s *Storage) View(ctx context.Context, fn func(kv.Tx) error) error {
+	return s.store.View(ctx, fn)
+}
+
+func (s *Storage) Update(ctx context.Context, fn func(kv.Tx) error) error {
+	return s.store.Update(ctx, fn)
+}
+
+// GetSecret Returns the value of a secret
+func (s *Storage) GetSecret(ctx context.Context, tx kv.Tx, orgID influxdb.ID, k string) (string, error) {
+	key, err := encodeSecretKey(orgID, k)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := tx.Bucket(secretBucket)
+	if err != nil {
+		return "", err
+	}
+
+	val, err := b.Get(key)
+	if kv.IsNotFound(err) {
+		return "", &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  influxdb.ErrSecretNotFound,
+		}
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	v, err := decodeSecretValue(val)
+	if err != nil {
+		return "", err
+	}
+
+	return v, nil
+}
+
+// ListSecrets returns a list of secret keys
+func (s *Storage) ListSecret(ctx context.Context, tx kv.Tx, orgID influxdb.ID) ([]string, error) {
+	b, err := tx.Bucket(secretBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	prefix, err := orgID.Encode()
+	if err != nil {
+		return nil, err
+	}
+
+	cur, err := b.ForwardCursor(prefix, kv.WithCursorPrefix(prefix))
+	if err != nil {
+		return nil, err
+	}
+
+	keys := []string{}
+
+	for k, _ := cur.Next(); k != nil; k, _ = cur.Next() {
+
+		id, key, err := decodeSecretKey(k)
+		if err != nil {
+			return nil, err
+		}
+
+		if id != orgID {
+			// We've reached the end of the keyspace for the provided orgID
+			break
+		}
+
+		keys = append(keys, key)
+	}
+
```

Comment:
You'll need to err check and close your cursor here.

Everyone does this, makes me think about the UX of cursors.

On that actually there is a `kv.Walk` that consumes a cursor and lets you provide a visit func. That does the cursor cleanup for you.

Reply:
oo thanks! I will fix this.